pkg/cni/plugin_test.go

```diff
@@ -0,0 +1,195 @@
+	var originalNS ns.NetNS
+	var targetNs ns.NetNS
+	var lowerDevice netlink.Link
+	var macvtapInterface netlink.Link
+	var deviceID string
+	var stdInArgs string
+
+	BeforeEach(func() {
+		var err error
+		originalNS, err = testutils.NewNS()
+		Expect(err).NotTo(HaveOccurred())
+		targetNs, err = testutils.NewNS()
+		Expect(err).NotTo(HaveOccurred())
+
+		err = originalNS.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+
+			// create lower device
+			err = netlink.LinkAdd(&netlink.Dummy{
+				LinkAttrs: netlink.LinkAttrs{
+					Name: LOWER_DEVICE,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+			lowerDevice, err = netlink.LinkByName(LOWER_DEVICE)
+			Expect(err).NotTo(HaveOccurred())
+
+			// create macvtap on top of lower device
+			macvtapIfaceName := "macvtap0"
+			_, err = util.CreateMacvtap(macvtapIfaceName, LOWER_DEVICE, "bridge")
+			Expect(err).NotTo(HaveOccurred())
+			macvtapInterface, err = netlink.LinkByName(macvtapIfaceName)
+			Expect(err).NotTo(HaveOccurred())
+			deviceID = macvtapIfaceName
+
+			stdInArgs = fmt.Sprintf(`{
+				"cniVersion": "0.3.1",
+				"name": "mynet",
+				"type": "macvtap",
+				"deviceID": "%s"
+			}`, macvtapIfaceName)
+			return nil
+		})
+
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		// cleanup the original namespace
+		originalNS.Do(func(ns.NetNS) error {
+			return netlink.LinkDel(lowerDevice)
+		})
+		Expect(originalNS.Close()).To(Succeed())
+		Expect(testutils.UnmountNS(originalNS)).To(Succeed())
+
+		// cleanup the target namespace
+		targetNs.Do(func(ns.NetNS) error {
+			return netlink.LinkDel(macvtapInterface)
+		})
+		Expect(targetNs.Close()).To(Succeed())
+		Expect(testutils.UnmountNS(targetNs)).To(Succeed())
+	})
+
+	It("imports a macvtap interface into the target netns", func() {
+		args := &skel.CmdArgs{
+			ContainerID: "dummy",
+			Netns:       targetNs.Path(),
+			IfName:      deviceID,
+			StdinData:   []byte(stdInArgs),
+		}
+
+		// import existing interface to target namespace
+		originalNS.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+
+			_, _, err := testutils.CmdAdd(args.Netns, args.ContainerID, args.IfName, args.StdinData, func() error { return cni.CmdAdd(args) })
+			Expect(err).NotTo(HaveOccurred())
+			return nil
+		})
+
+		// confirm macvtap is available on target namespace
+		targetNs.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+			_, err := netlink.LinkByName(deviceID)
+			Expect(err).NotTo(HaveOccurred())
+			return nil
+		})
+	})
+
+	It("removes a macvtap interface", func() {
+		args := &skel.CmdArgs{
+			ContainerID: "dummy",
+			Netns:       targetNs.Path(),
+			IfName:      deviceID,
+			StdinData:   []byte(stdInArgs),
+		}
+
+		// import existing interface to target namespace
+		originalNS.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+
+			_, _, err := testutils.CmdAdd(args.Netns, args.ContainerID, args.IfName, args.StdinData, func() error { return cni.CmdAdd(args) })
```

Comment:
Here the setup would fit into `BeforeEach` (it is not the subject of testing), while in `It`, we would have only the removal.

Reply:
Done